mail-storage/src/main/java/mail/storage/service/MailStorageServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+package mail.storage.service;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import mail.storage.domain.Message;
+import mail.storage.domain.MessageType;
+import mail.storage.dto.DateRangeDto;
+import mail.storage.dto.MessageDto;
+import mail.storage.dto.UpdateMessageDto;
+import mail.storage.exception.DraftMessageException;
+import mail.storage.exception.MessageWithNumberAlreadyExists;
+import mail.storage.exception.MessageWithNumberNotFound;
+import mail.storage.repository.MessageRepository;
+import mail.storage.util.MessageUtils;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+import static mail.storage.util.MessageUtils.getMessageFromDto;
+
+
+@Service
+@RequiredArgsConstructor
+@Log4j2
+@Getter
+public class MailStorageServiceImpl implements MailStorageService {
+    private final MessageRepository repository;
+
+    public void addMessage(MessageDto messageDto) throws MessageWithNumberAlreadyExists {
+        if (numberOfMessageAlreadyExists(messageDto)) {
+            throw new MessageWithNumberAlreadyExists(String.format("Message with number %d already exists", messageDto.getNumber()));
+        } else {
+            repository.save(getMessageFromDto(messageDto));
+        }
+    }
+
+    private boolean numberOfMessageAlreadyExists(MessageDto messageDto) {
+        return repository.findByNumber(messageDto.getNumber()).isPresent();
```

Comment:
.isPresent() in Optional is ugly -> rework addMsg - you can Optional.map orElseThrow etc

Reply:
Fixed.